### ISSUE TITLE
Warning when initializing dataclasses with arguments under python 3.6 using the dataclasses backport

### ISSUE THREAD

**Issue by Delwin9999** [microsoft/pyright#2003]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
When using the dataclasses backport under python 3.6 pyright does not understand dataclasses and gives warnings when trying to initialize them with arguments.

**To Reproduce**
Create a python file with the following:
```python
from dataclasses import dataclass


@dataclass
class MyDataclass:
    foo: str


MyDataclass(foo="bar")
```

**Expected behavior**
There should be no pyright warnings from the above code.

**Screenshots or Code**
Screenshots of VS Code using python 3.6 (and dataclasses 0.8) and python 3.8 on the same code above.
![pyright_dataclass_failing](https://user-images.githubusercontent.com/817061/122488184-cb1fd280-cf91-11eb-8155-df8509037e3f.jpg)
![pyright_dataclass_working](https://user-images.githubusercontent.com/817061/122488187-cd822c80-cf91-11eb-89a0-4e132c008405.jpg)

**VS Code extension or command-line**
Pyright VS Code extension 1.1.150 - also pictured in the above screenshots.

**Additional context**
Initially referenced in #1433 but not reproducible because the difference in python versions was not taken into account.

**Comment by erictraut** [comment]
Thanks for the bug report. This previously worked fine, but a change over the past few months within the typeshed stubs caused it to break. I've updated the type evaluator to handle the change in the stubs.

**Comment by erictraut** [comment]
This is included in pyright 1.1.151, which I just published. It will also be included in the next release of pylance.

**Comment by rom4kov** [comment]
I get the same error with python 3.12 and pyright 1.1.368
![screenshot_2024-06-20-142539](https://github.com/microsoft/pyright/assets/102757050/c8a3e893-6f4f-484d-be57-51e630425498)

**Comment by rom4kov** [comment]
I solved it by adding an __init__ function to the class (so obvious):

`class Book(db.Model):
    id: Mapped[int] = mapped_column(Integer(), primary_key=True)
    title: Mapped[str] = mapped_column(String(250), nullable=False, unique=True)
    author: Mapped[str] = mapped_column(String(250), nullable=False)
    rating: Mapped[float] = mapped_column(Float(), nullable=False)

    def __init__(self, id: int, title: str, author: str, rating: float):
        self.id = id
        self.title = title
        self.author = author
        self.rating = rating`
        
But strangely, I didn't get any error in PyCharm without the __init__ function.

**Comment by rom4kov** [comment]
I solved it by adding an init function to the class (so obvious):

`
class Book(db.Model):
id: Mapped[int] = mapped_column(Integer(), primary_key=True)
title: Mapped[str] = mapped_column(String(250), nullable=False, unique=True)
author: Mapped[str] = mapped_column(String(250), nullable=False)
rating: Mapped[float] = mapped_column(Float(), nullable=False)

def __init__(self, id: int, title: str, author: str, rating: float):
    self.id = id
    self.title = title
    self.author = author
    self.rating = rating
`

But strangely, I didn't get any error in PyCharm without the init function.

**Comment by rom4kov** [comment]
I solved it by adding an init function to the class (so obvious):

`class Book(db.Model):
id: Mapped[int] = mapped_column(Integer(), primary_key=True)
title: Mapped[str] = mapped_column(String(250), nullable=False, unique=True)
author: Mapped[str] = mapped_column(String(250), nullable=False)
rating: Mapped[float] = mapped_column(Float(), nullable=False)

def init(self, id: int, title: str, author: str, rating: float):
self.id = id
self.title = title
self.author = author
self.rating = rating`

But strangely, I didn't get any error in PyCharm without the init function.

**Comment by rom4kov** [comment]
I solved it by adding an init function to the class.

```
class Book(db.Model):
    id: Mapped[int] = mapped_column(Integer(), primary_key=True)
    title: Mapped[str] = mapped_column(String(250), nullable=False, unique=True)
    author: Mapped[str] = mapped_column(String(250), nullable=False)
    rating: Mapped[float] = mapped_column(Float(), nullable=False)

    def init(self, id: int, title: str, author: str, rating: float):
        self.id = id
        self.title = title
        self.author = author
        self.rating = rating
```

But strangely, I didn't get any error in PyCharm without the init function.

**Comment by untitaker** [comment]
what does this have to do with dataclasses? it seems to be an issue with sqlalchemy stubs. please don't necrobump old issues.

On Thu, Jun 20, 2024, at 15:06, Roman wrote:
> 
> 
> I solved it by adding an *init* function to the class (so obvious):
> 
> 
> `class Book(db.Model):
> id: Mapped[int] = mapped_column(Integer(), primary_key=True)
> title: Mapped[str] = mapped_column(String(250), nullable=False, unique=True)
> author: Mapped[str] = mapped_column(String(250), nullable=False)
> rating: Mapped[float] = mapped_column(Float(), nullable=False)
> 
> `def __init__(self, id: int, title: str, author: str, rating: float):
>     self.id = id
>     self.title = title
>     self.author = author
>     self.rating = rating`
`

**Comment by rom4kov** [comment]
You're right, of course, my bad. Should I delete the posts, then?

**Comment by rom4kov** [comment]
I've deleted my previous comment as it was off-topic. Apologies for any confusion.
